File: nachna/lib/models/rewards.dart
import 'package:json_annotation/json_annotation.dart';

part 'rewards.g.dart';

enum RewardSourceEnum {
  @JsonValue('referral')
  referral,
  @JsonValue('cashback')
  cashback,
  @JsonValue('welcome_bonus')
  welcomeBonus,
  @JsonValue('special_promotion')
  specialPromotion,
  @JsonValue('workshop_completion')
  workshopCompletion,
  @JsonValue('admin_bonus')
  adminBonus,
}

enum RewardTransactionTypeEnum {
  @JsonValue('credit')
  credit,
  @JsonValue('debit')
  debit,
}

enum RewardTransactionStatusEnum {
  @JsonValue('pending')
  pending,
  @JsonValue('completed')
  completed,
  @JsonValue('failed')
  failed,
  @JsonValue('cancelled')
  cancelled,
}

@JsonSerializable()
class RewardBalance {
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'total_balance')
  final double totalBalance;
  @JsonKey(name: 'available_balance')
  final double availableBalance;
  @JsonKey(name: 'lifetime_earned')
  final double lifetimeEarned;
  @JsonKey(name: 'lifetime_redeemed')
  final double lifetimeRedeemed;
  @JsonKey(name: 'redemption_cap_per_workshop')
  final double redemptionCapPerWorkshop;

  RewardBalance({
    required this.userId,
    required this.totalBalance,
    required this.availableBalance,
    required this.lifetimeEarned,
    required this.lifetimeRedeemed,
    required this.redemptionCapPerWorkshop,
  });

  factory RewardBalance.fromJson(Map<String, dynamic> json) => _$RewardBalanceFromJson(json);
  Map<String, dynamic> toJson() => _$RewardBalanceToJson(this);

  // Helper getters
  String get formattedTotalBalance => '₹${totalBalance.toStringAsFixed(0)}';
  String get formattedAvailableBalance => '₹${availableBalance.toStringAsFixed(0)}';
  String get formattedLifetimeEarned => '₹${lifetimeEarned.toStringAsFixed(0)}';
  String get formattedLifetimeRedeemed => '₹${lifetimeRedeemed.toStringAsFixed(0)}';
  String get formattedRedemptionCap => '₹${redemptionCapPerWorkshop.toStringAsFixed(0)}';
}

@JsonSerializable()
class RewardTransaction {
  @JsonKey(name: 'transaction_id')
  final String transactionId;
  @JsonKey(name: 'transaction_type')
  final RewardTransactionTypeEnum transactionType;
  final double amount;
  final String source;
  final String status;
  final String description;
  @JsonKey(name: 'reference_id')
  final String? referenceId;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'processed_at')
  final DateTime? processedAt;

  RewardTransaction({
    required this.transactionId,
    required this.transactionType,
    required this.amount,
    required this.source,
    required this.status,
    required this.description,
    this.referenceId,
    required this.createdAt,
    this.processedAt,
  });

  factory RewardTransaction.fromJson(Map<String, dynamic> json) => _$RewardTransactionFromJson(json);
  Map<String, dynamic> toJson() => _$RewardTransactionToJson(this);

  // Helper getters
  String get formattedAmount => '₹${amount.toStringAsFixed(0)}';
  String get formattedDate => '${createdAt.day}/${createdAt.month}/${createdAt.year}';
  bool get isCredit => transactionType == RewardTransactionTypeEnum.credit;
  bool get isDebit => transactionType == RewardTransactionTypeEnum.debit;
  
  String get displayTitle {
    switch (source) {
      case 'referral':
        return 'Referral Bonus';
      case 'cashback':
        return isDebit ? 'Workshop Discount' : 'Cashback Earned';
      case 'welcome_bonus':
        return 'Welcome Bonus';
      case 'special_promotion':
        return 'Special Promotion';
      case 'workshop_completion':
        return 'Workshop Completion';
      case 'admin_bonus':
        return 'Admin Bonus';
      default:
        return 'Reward Transaction';
    }
  }
}

@JsonSerializable()
class RewardTransactionList {
  final List<RewardTransaction> transactions;
  @JsonKey(name: 'total_count')
  final int totalCount;
  final int page;
  @JsonKey(name: 'page_size')
  final int pageSize;

  RewardTransactionList({
    required this.transactions,
    required this.totalCount,
    required this.page,
    required this.pageSize,
  });

  factory RewardTransactionList.fromJson(Map<String, dynamic> json) => _$RewardTransactionListFromJson(json);
  Map<String, dynamic> toJson() => _$RewardTransactionListToJson(this);
}

@JsonSerializable()
class RewardSummary {
  final RewardBalance balance;
  @JsonKey(name: 'recent_transactions')
  final List<RewardTransaction> recentTransactions;
  @JsonKey(name: 'total_savings')
  final double totalSavings;
  @JsonKey(name: 'redemption_history_count')
  final int redemptionHistoryCount;

  RewardSummary({
    required this.balance,
    required this.recentTransactions,
    required this.totalSavings,
    required this.redemptionHistoryCount,
  });

  factory RewardSummary.fromJson(Map<String, dynamic> json) => _$RewardSummaryFromJson(json);
  Map<String, dynamic> toJson() => _$RewardSummaryToJson(this);

  // Helper getters
  String get formattedTotalSavings => '₹${totalSavings.toStringAsFixed(0)}';
}

@JsonSerializable()
class WorkshopRedemptionInfo {
  @JsonKey(name: 'workshop_uuid')
  final String workshopUuid;
  @JsonKey(name: 'workshop_title')
  final String workshopTitle;
  @JsonKey(name: 'original_amount')
  final double originalAmount;
  @JsonKey(name: 'max_redeemable_points')
  final double maxRedeemablePoints;
  @JsonKey(name: 'max_discount_amount')
  final double maxDiscountAmount;
  @JsonKey(name: 'user_available_balance')
  final double userAvailableBalance;
  @JsonKey(name: 'recommended_redemption')
  final double recommendedRedemption;

  WorkshopRedemptionInfo({
    required this.workshopUuid,
    required this.workshopTitle,
    required this.originalAmount,
    required this.maxRedeemablePoints,
    required this.maxDiscountAmount,
    required this.userAvailableBalance,
    required this.recommendedRedemption,
  });

  factory WorkshopRedemptionInfo.fromJson(Map<String, dynamic> json) => _$WorkshopRedemptionInfoFromJson(json);
  Map<String, dynamic> toJson() => _$WorkshopRedemptionInfoToJson(this);

  // Helper getters
  String get formattedOriginalAmount => '₹${originalAmount.toStringAsFixed(0)}';
  String get formattedMaxDiscount => '₹${maxDiscountAmount.toStringAsFixed(0)}';
  String get formattedAvailableBalance => '₹${userAvailableBalance.toStringAsFixed(0)}';
  String get formattedRecommendedRedemption => '₹${recommendedRedemption.toStringAsFixed(0)}';
}

@JsonSerializable()
class RedemptionCalculation {
  @JsonKey(name: 'workshop_info')
  final WorkshopRedemptionInfo workshopInfo;
  @JsonKey(name: 'exchange_rate')
  final double exchangeRate;
  @JsonKey(name: 'can_redeem')
  final bool canRedeem;
  final String? message;

  RedemptionCalculation({
    required this.workshopInfo,
    required this.exchangeRate,
    required this.canRedeem,
    this.message,
  });

  factory RedemptionCalculation.fromJson(Map<String, dynamic> json) => _$RedemptionCalculationFromJson(json);
  Map<String, dynamic> toJson() => _$RedemptionCalculationToJson(this);
}

@JsonSerializable()
class RewardRedemptionRequest {
  @JsonKey(name: 'workshop_uuid')
  final String workshopUuid;
  @JsonKey(name: 'points_to_redeem')
  final double pointsToRedeem;
  @JsonKey(name: 'order_amount')
  final double orderAmount;

  RewardRedemptionRequest({
    required this.workshopUuid,
    required this.pointsToRedeem,
    required this.orderAmount,
  });

  factory RewardRedemptionRequest.fromJson(Map<String, dynamic> json) => _$RewardRedemptionRequestFromJson(json);
  Map<String, dynamic> toJson() => _$RewardRedemptionRequestToJson(this);
}

@JsonSerializable()
class RewardRedemption {
  @JsonKey(name: 'redemption_id')
  final String redemptionId;
  @JsonKey(name: 'points_redeemed')
  final double pointsRedeemed;
  @JsonKey(name: 'discount_amount')
  final double discountAmount;
  @JsonKey(name: 'original_amount')
  final double originalAmount;
  @JsonKey(name: 'final_amount')
  final double finalAmount;
  @JsonKey(name: 'savings_percentage')
  final double savingsPercentage;
  final String status;

  RewardRedemption({
    required this.redemptionId,
    required this.pointsRedeemed,
    required this.discountAmount,
    required this.originalAmount,
    required this.finalAmount,
    required this.savingsPercentage,
    required this.status,
  });

  factory RewardRedemption.fromJson(Map<String, dynamic> json) => _$RewardRedemptionFromJson(json);
  Map<String, dynamic> toJson() => _$RewardRedemptionToJson(this);

  // Helper getters
  String get formattedPointsRedeemed => '${pointsRedeemed.toStringAsFixed(0)} pts';
  String get formattedDiscountAmount => '₹${discountAmount.toStringAsFixed(0)}';
  String get formattedOriginalAmount => '₹${originalAmount.toStringAsFixed(0)}';
  String get formattedFinalAmount => '₹${finalAmount.toStringAsFixed(0)}';
  String get formattedSavingsPercentage => '${savingsPercentage.toStringAsFixed(1)}%';
}

File: app/models/rewards.py
"""Rewards system models for Pydantic validation and API responses."""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from enum import Enum


class RewardSourceEnum(str, Enum):
    """Types of reward sources."""
    REFERRAL = "referral"
    CASHBACK = "cashback"
    WELCOME_BONUS = "welcome_bonus"
    SPECIAL_PROMOTION = "special_promotion"
    WORKSHOP_COMPLETION = "workshop_completion"
    ADMIN_BONUS = "admin_bonus"
    REFUND = "refund"


class RewardTransactionTypeEnum(str, Enum):
    """Types of reward transactions."""
    CREDIT = "credit"  # Points earned
    DEBIT = "debit"    # Points redeemed


class RewardTransactionStatusEnum(str, Enum):
    """Status of reward transactions."""
    PENDING = "pending"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


# Database Models
class RewardWallet(BaseModel):
    """User's reward wallet."""
    user_id: str
    total_balance: float = Field(default=0.0, description="Total reward points balance")
    available_balance: float = Field(default=0.0, description="Available points for redemption")
    lifetime_earned: float = Field(default=0.0, description="Total points earned lifetime")
    lifetime_redeemed: float = Field(default=0.0, description="Total points redeemed lifetime")
    created_at: datetime
    updated_at: datetime


class RewardTransaction(BaseModel):
    """Individual reward transaction."""
    transaction_id: str
    user_id: str
    transaction_type: RewardTransactionTypeEnum
    amount: float = Field(description="Amount of reward points")
    source: RewardSourceEnum
    status: RewardTransactionStatusEnum
    description: str = Field(description="Human readable description")
    reference_id: Optional[str] = Field(None, description="Reference to order, referral, etc.")
    metadata: Optional[dict] = Field(None, description="Additional transaction data")
    created_at: datetime
    processed_at: Optional[datetime] = None


class RewardRedemption(BaseModel):
    """Reward redemption for workshop booking."""
    redemption_id: str
    user_id: str
    order_id: str
    workshop_uuid: str
    points_redeemed: float
    discount_amount: float = Field(description="Monetary discount equivalent")
    original_amount: float
    final_amount: float
    status: RewardTransactionStatusEnum
    created_at: datetime
    processed_at: Optional[datetime] = None


# API Request Models
class RewardRedemptionRequest(BaseModel):
    """Request to redeem rewards for workshop booking."""
    workshop_uuid: str
    points_to_redeem: float = Field(gt=0, description="Points to redeem (must be positive)")
    order_amount: float = Field(gt=0, description="Original order amount")


class RewardBalanceResponse(BaseModel):
    """Response with user's reward balance information."""
    user_id: str
    total_balance: float
    available_balance: float
    lifetime_earned: float
    lifetime_redeemed: float
    redemption_cap_per_workshop: float = Field(description="Maximum points redeemable per workshop")


class RewardTransactionResponse(BaseModel):
    """Response for reward transaction."""
    transaction_id: str
    transaction_type: RewardTransactionTypeEnum
    amount: float
    source: RewardSourceEnum
    status: RewardTransactionStatusEnum
    description: str
    reference_id: Optional[str]
    created_at: datetime
    processed_at: Optional[datetime]


class RewardTransactionListResponse(BaseModel):
    """Response with list of reward transactions."""
    transactions: List[RewardTransactionResponse]
    total_count: int
    page: int
    page_size: int


class RewardRedemptionResponse(BaseModel):
    """Response for reward redemption."""
    redemption_id: str
    points_redeemed: float
    discount_amount: float
    original_amount: float
    final_amount: float
    savings_percentage: float = Field(description="Percentage saved")
    status: RewardTransactionStatusEnum


class RewardSummaryResponse(BaseModel):
    """Summary response for rewards center."""
    balance: RewardBalanceResponse
    recent_transactions: List[RewardTransactionResponse]
    total_savings: float = Field(description="Total money saved through rewards")
    redemption_history_count: int


# Workshop Redemption Models
class WorkshopRedemptionInfo(BaseModel):
    """Information about reward redemption for a specific workshop."""
    workshop_uuid: str
    workshop_title: str
    original_amount: float
    max_redeemable_points: float
    max_discount_amount: float
    user_available_balance: float
    recommended_redemption: float = Field(description="Recommended points to redeem")


class RedemptionCalculationRequest(BaseModel):
    """Request to calculate redemption for workshop."""
    workshop_uuid: str
    workshop_amount: float


class RedemptionCalculationResponse(BaseModel):
    """Response with redemption calculation."""
    workshop_info: WorkshopRedemptionInfo
    exchange_rate: float = Field(description="Points to currency exchange rate (e.g., 1 point = 1 rupee)")
    can_redeem: bool
    message: Optional[str] = Field(description="Message if redemption not possible")
